hazelcast/include/hazelcast/client/ICompletableFuture.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef HAZELCAST_CLIENT_ICOMPLETABLEFUTURE_H_
+#define HAZELCAST_CLIENT_ICOMPLETABLEFUTURE_H_
+
+#include <boost/shared_ptr.hpp>
+
+#include "hazelcast/client/IFuture.h"
+#include "hazelcast/client/ExecutionCallback.h"
+#include "hazelcast/util/Executor.h"
+
+using namespace hazelcast::client;
+using namespace hazelcast::util;
+
+namespace hazelcast {
+    namespace client {
+
+        /**
+         * A Future where one can asynchronously listen on completion. This functionality is needed for the
+         * reactive programming model.
+         *
+         * @param <V> The result type returned by this Future's {@code get} method
+         */
+        template<typename V>
+        class ICompletableFuture : public IFuture<V> {
```

Comment:
Missing virtual destructor

Reply:
The derived abstract class (`AbstractInvocationFuture`) has the virtual destructor, hence I thought it was not needed at this interface. Do we also need it at the interface?

Reply:
The rule I use is:
If a class will be used polymorphically it needs a virtual destructor. 

More specifically, If a derived class is deleted via the base class,  and the base class does not have a virtual destructor, the behavior is undefined.